### ISSUE TITLE
[RFC] Optimizing CUDA+AArch64 builds using LLVM-BOLT

### ISSUE THREAD

**Issue by lakshayg** [pytorch/pytorch#188602]
[LLVM BOLT](https://github.com/llvm/llvm-project/blob/main/bolt/README.md) is a post-link optimizer that optimize's application's code layout based on execution profile gathered by perf. It has been used to optimize NVIDIA's PyTorch containers since [26.04](https://github.com/llvm/llvm-project/blob/main/bolt/README.md). This PR brings that optimization into the PyTorch repo.

LLVM BOLT works in two phases, first a profile is collected using an unoptimized binary and then it is used to optimize the binary.

The list of tasks that need to be done:

- Basic tooling
  - [x] Integrate LLVM-BOLT into the build system
    - https://github.com/pytorch/pytorch/pull/186245
  - [x] Make linker script work with LLD
    - https://github.com/pytorch/pytorch/pull/186245
    - https://github.com/pytorch/pytorch/pull/189891
- Enablement in CI
  - [x] Ensure PyTorch CI has the right BOLT version to apply the optimization
    - https://github.com/pytorch/pytorch/pull/190807
  - [ ] Install BOLT in all containers, use only on cuda+aarch64, comment from @malfet 
    - https://github.com/pytorch/pytorch/pull/190807
  - [ ] Align the toolchain used to collect fdata with the CI toolchain (python, cuda, bolt, compiler flags, linker flags etc)
    - https://github.com/pytorch/pytorch/pull/195221
- Lifecycle
  - [ ] Add a profile staleness metric script which guides profile updates
    - https://github.com/pytorch/pytorch/pull/195026
  - [ ] Enable profile collection in PyTorch CI
- Other tasks
  - [ ] Remove linker script optimization
  - [ ] Consider using `lld` in CI to avoid build slowdown due to `--emit-relocs` in `RelWithDebInfo` builds
  - [ ] Remove libtorch.yaml from torch.tar.zst
    - See: https://github.com/pytorch/pytorch/pull/194941
  - [ ] Address @aaupov's comments on https://github.com/pytorch/pytorch/pull/186245
    - https://github.com/pytorch/pytorch/pull/195475
  - [ ] Improve bring-your-own-fdata workflow

cc @malfet @ptrblck @msaroufim @eqy @tinglvv @nWEIdia @pytorch/pytorch-dev-infra @snadampal @milpuz01 @aditew01 @nikhil-arm @fadara01 @robert-hardwick

**Comment by drisspg** [comment]
Can you add the real throuhgput improvement we expect from this?

**Comment by drisspg** [comment]
Notes from composability: 


1. How do we know which workloads this will help and which ones this will hurt?
2. Do we have a sense of the impact to compile time?
3.  Whats the difference between Bolt, LTO, and PGO ? Which are currently applied to pytorch? 
	1. April 2024 linker script added and never changed -> probably stale
4. How is this not going to end up stale. 


Utopia -> always run bolt profiler on important hardware on representative workloads and produce faster/better/stronger wheels. 

Requirements:
1.  We need details on how when where the profiles were generated. If in the limit we wanted this running at nightly build time.


Tangential: how do we measure the old linker script staleness

**Comment by lakshayg** [comment]
@drisspg here are the performance numbers for BOLT profile data collected in Jan 2026: https://gist.github.com/lakshayg/0588e8cb7a270c2edcaac57558704f57

> How do we know which workloads this will help and which ones this will hurt?

I don't think we can determine this ahead of time. As long as a workload is adequately represented our profiling set, it should see an improvement. Call stacks that are common in profiled workloads in general will see an improvement.

> Do we have a sense of the impact to compile time?

All the results below are for RelWithDebInfo build configuration.

I profiled the build and noticed a few things:
- The compile time remains mostly unchanged
- The end to end time to produce the build artifacts goes from ~15min to ~33min on GB200
  - Almost all increase comes from the increase in link time
  - The link time for libtorch_cpu increases a lot. 33 sec -> 21 min. This is likely because of --emit-relocs
  - The llvm-bolt execution itself takes ~5min

**Update Jul 14, 2026**:
Good news: The linker overhead goes away if I switch from `ld` to `lld`.
Bad news: The linker script optimization requires `ld`. We either need to update the script to support `lld` or disable linker script altogether when BOLT optimization is enabled.

> Whats the difference between Bolt, LTO, and PGO ? Which are currently applied to pytorch?

I can share my mental model of the three optimizations you mentioned:
1. LTO - This is about optimizing across the boundaries of a binary. Imagine if all the code was put into a single file. This helps with inlining and dead code elimination.
2. PGO - You compile the program, profile it, and finally compile again using the profile. This can be pretty expensive.
3. BOLT - It optimizes the compiled binary itself, no recompilation is needed.

There's a nice article/slides/video about optimizing the linux kernel using all 3 here: https://lwn.net/Articles/993828/

I don't know which ones are applied to PyTorch but I'm sure PGO is not applied.

> How is this not going to end up stale

It will end up going stale. The plan is to add a staleness metric script and monitor the binaries using that. Once the staleness exceeds an agreed upon threshold, we can update the profile data.

**Comment by paschalis-mpeis** [comment]
Hey folks,

Great to see PyTorch considering BOLT. Please let us know if you run into any issues; we'd be happy to help.

For background on the different PGO/FDO variants, this is a useful overview **[The many faces of LLVM PGO and FDO](https://aaupov.github.io/blog/2023/07/09/pgo)**.

I also made a simplified video tutorial that explains **[LTO vs PGO vs BOLT](https://www.youtube.com/watch?v=KdHtOMc5_c8)** (first 25 minutes), including how to check whether an application is a good BOLT candidate and how to verify that BOLT has made a difference.

**Comment by lakshayg** [comment]
@drisspg is there any other info you think would be useful for the maintainers/reviewers?

**Comment by drisspg** [comment]
Hey, @lakshayg  sorry for the late reply.  All makes sense I think we are in agreement that we can start working on this

**Comment by aaupov** [comment]
Hi @drisspg, couple of comments from Meta/BOLT perspective:

> Notes from composability:
> 
> 1. How do we know which workloads this will help and which ones this will hurt?

In general, the workloads where compiler optimizations can help is where CPU overhead is throttling GPU utilization. More specifically, training workloads with large CPU code footprint where CUDA kernel launch is on the critical path. Compiler optimizations are not expected to cause performance regressions, worst case they won't affect GPU utilization/QPS. However, compiler optimizations have operational costs (maintaining profiles) and side effects such as build time and binary size.

> 3. Do we have a sense of the impact to compile time?

Applying BOLT to small- or medium-sized binaries/shared objects is very fast, usually taking less than a minute. So in the build pipeline, BOLT incurs marginal overhead on top of compilation and linking. 

However, profile collection and conversion are offline but necessary steps that must be accounted for. Profile collection is dominated by workload setup to get representative profile (warmup etc). Sampled profile itself can be collected in a very short time (~minutes is usually sufficient), and profile conversion (perf2bolt) is as fast as optimization. Profiling using instrumented binary incurs larger overheads, similar to instrumentation PGO, but doesn't require any profile conversion.

> 5. Whats the difference between Bolt, LTO, and PGO ? Which are currently applied to pytorch?

As mentioned above, LTO, PGO, BOLT occupy different places in optimal compilation pipeline.

* LTO is a way to make compiler optimizations more effective by exposing whole program to cross-CU optimizations. LTO doesn't require the profile, hence it's the lowest hanging fruit/highest ROI optimization.
* Compiler PGO is usually the biggest lever where profile information drives better optimizations (mainly through aggressive but accurate inlining), which is synergistic with LTO. There are different forms of PGO depending on what kind of profile is used: instrumentation (IPGO) or sampled (AutoFDO/Sampled PGO), where the profile gets attached (FE PGO, IR PGO, CSIR PGO, FS-AFDO, CSSPGO, ...).
* BOLT is a post-link optimizer, which has some advantages over compiler optimizations: 
1. profile matching is lossless (profiled binary = pre-bolt binary),
2. by virtue of being disassembler BOLT can optimize functions not reachable by compiler (from libraries or written in assembly), 
3. all while avoiding recompilation.

>    1. April 2024 linker script added and never changed -> probably stale

Linker script is perhaps the most robust form of PGO which doesn't require as precise/up-to-date profile information as PGO or BOLT. But still, linker script order file has to be kept in sync with source, otherwise it's best replaced by PGO or BOLT.

> 4. How is this not going to end up stale.
> 
> Utopia -> always run bolt profiler on important hardware on representative workloads and produce faster/better/stronger wheels.

> 
> Requirements:
> 
> 1. We need details on how when where the profiles were generated. If in the limit we wanted this running at nightly build time.
> 
> Tangential: how do we measure the old linker script staleness

BOLT has a way of reporting useful stats about existing binary layout using sampled profile as input data. If you run perf2bolt and add `-heatmap` flag it will produce a heatmap for visual inspection of how well did linker script group hot functions together. If you use a linker script injecting __hot_start and __hot_end marker symbols, BOLT can compute hot text utilization (% of hot text that was executed), and hotness (% of samples falling into hot text). Multiplying the two you get a single figure representing the quality of optimization profile wrt sampled profile – we call it _partition score_. And if sampled profile is derived from a wide range of production uses, meaning it's representative by construction, the partition score is a measure of _optimization_ profile representativeness.
